### PR TITLE
integrals : Fixed polytope_integrate for max_degree inputs

### LINKED ISSUE #19234: polytope_integrate and max_degree
Trying to build on the examples as `https://docs.sympy.org/latest/modules/integrals/integrals.html#sympy.integrals.intpoly.polytope_integrate:
```
polygon = Polygon(Point(0, 0), Point(0, 1), Point(1, 1), Point(1, 0))
polys = [1, x, y, x*y, x**2*y, x*y**2]
polytope_integrate(polygon, polys, max_degree=3)
```
works but
```
polytope_integrate(polygon, polys)
```
fails:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/61.426227578929124852/sympy/sympy/integrals/intpoly.py", line 136, in polytope_integrate
    return main_integrate(expr, facets, hp_params)
  File "/base/data/home/apps/s~sympy-live-hrd/61.426227578929124852/sympy/sympy/integrals/intpoly.py", line 306, in main_integrate
    dims, deg)
  File "/base/data/home/apps/s~sympy-live-hrd/61.426227578929124852/sympy/sympy/integrals/intpoly.py", line 462, in integration_reduction
    expr = _sympify(expr)
  File "/base/data/home/apps/s~sympy-live-hrd/61.426227578929124852/sympy/sympy/core/sympify.py", line 418, in _sympify
    return sympify(a, strict=True)
  File "/base/data/home/apps/s~sympy-live-hrd/61.426227578929124852/sympy/sympy/core/sympify.py", line 340, in sympify
    raise SympifyError(a)
SympifyError: SympifyError: [1, x, y, x*y, x**2*y, x*y**2]
```
Is that the expected behavior? Why? Thanks!

### PR DESCRIPTION
**Pull request by faze-geek**
#### References to other Issues or PRs

Fixes #19234

#### Brief description of what is fixed or changed
Maximum degree and list inputs were not handled correctly so code has been refactored.
Some incorrect things fixed are -
1. Only maximum possible degree in the list is a valid input which doesn't make sense for user input.
2. If maximum degree is not given ,all available degrees should be returned since max_degree is optional input.
3. Change in error message which was inappropriate.

**UPDATE**  - Now these issues are fixed for polynomials along with monomials too.

Current state of master -
```
>>> from sympy import *
>>> from sympy.integrals.intpoly import *
>>> polygon = Polygon(Point(0, 0), Point(0, 1), Point(1, 1), Point(1, 0))
>>> polys =  [ 1,x, y, x*y, x**2*y, x*y**2]
>>> polytope_integrate(polygon, polys,max_degree = 3)
{1: 1, x: 1/2, y: 1/2, x*y: 1/4, x**2*y: 1/6, x*y**2: 1/6}
>>> polytope_integrate(polygon, polys,max_degree = 2)   
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\kunni\sympy\sympy\sympy\integrals\intpoly.py", line 126, in polytope_integrate
    integral_value += result_dict[m] * coeff
KeyError: x**2*y
>>> polytope_integrate(polygon, polys)   
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\kunni\sympy\sympy\sympy\integrals\intpoly.py", line 134, in polytope_integrate
    return main_integrate(expr, facets, hp_params)
  File "C:\Users\kunni\sympy\sympy\sympy\integrals\intpoly.py", line 302, in main_integrate
    value_over_boundary = integration_reduction(facets,
  File "C:\Users\kunni\sympy\sympy\sympy\integrals\intpoly.py", line 476, in integration_reduction
    expr = _sympify(expr)
  File "C:\Users\kunni\sympy\sympy\sympy\core\sympify.py", line 528, in _sympify
    return sympify(a, strict=True)
  File "C:\Users\kunni\sympy\sympy\sympy\core\sympify.py", line 449, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: [1, x, y, x*y, x**2*y, x*y**2]
```

After the change -
```
>>> from sympy import *
>>> from sympy.integrals.intpoly import *
>>> polygon = Polygon(Point(0, 0), Point(0, 1), Point(1, 1), Point(1, 0))
>>> polys =  [ 1,x, y, x*y, x**2*y, x*y**2]
>>> polytope_integrate(polygon, polys)
{1: 1, x: 1/2, y: 1/2, x*y: 1/4, x**2*y: 1/6, x*y**2: 1/6}
>>> polytope_integrate(polygon, polys,max_degree=2)
{1: 1, x: 1/2, y: 1/2, x*y: 1/4}
>>> polytope_integrate(polygon, polys,max_degree =1)
{1: 1, x: 1/2, y: 1/2}
```

#### Other comments
Some more tests could be added once pull request is approved .
Online documentation on this is very thin ,once this pr gets merged someone should improve it in a new pull request .
There are similar issues encountered with 3D polygons like cube which could be fixed in a future pr.
#### Release Notes

* integrals
  * Fixed polytype_integrate for max_degree inputs

### CODE REVIEW

**Comment by faze-geek** [comment]
Ping @smichr @oscargus for review . Timewise there is almost null difference between old code and refactored code but current code makes quite a few corrections in results , so I'll appreciate any suggestions/improvements .Thanks !!

File: sympy/integrals/intpoly.py
Comment (oscargus) [inline sympy/integrals/intpoly.py:19]: Import from sympy.poly (at least, possibly another level down).

File: sympy/integrals/intpoly.py
Comment (oscargus) [inline sympy/integrals/intpoly.py:307]: Try to factor out this code into a (private) function and use here and for the inner loop below.

**Comment by oscargus** [comment]
Some comments about the code quality. I have no idea about the mathematics of stuff here.

**Comment by faze-geek** [comment]
> Some comments about the code quality. I have no idea about the mathematics of stuff here.

Yeah I'll make the minor fixes, asked for a review since you had helped in improving previous pr's related to this .Thanks anyways :)

**Comment by faze-geek** [comment]
Ping @oscarbenjamin for review .

File: sympy/integrals/intpoly.py
Comment (faze-geek) [reply sympy/integrals/intpoly.py:327]: Yeah I have done this , hope that the name used is suitable !

**Comment by faze-geek** [comment]
CC @ArifAhmed1995 since you and @certik had collaborated for the code in `intpoly ` could you comment on these code improvements .Thanks .

File: sympy/integrals/intpoly.py
Comment (oscargus) [reply sympy/integrals/intpoly.py:325]: Looks good! Thanks! (Good that you used a leading underscore.)

File: sympy/integrals/intpoly.py
Comment (oscargus) [inline sympy/integrals/intpoly.py:327]:
```suggestion
            return _polynomial_integrate(polynomials, facets, hp_params)
```

File: sympy/integrals/intpoly.py
Comment (oscargus) [inline sympy/integrals/intpoly.py:335]:
```suggestion
            return {exp: _polynomial_integrate(decompose(exp), facets, hp_params) for exp in expr}
```
is a bit more compact/pythonic, but no big deal.

Comment (oscargus) [reply sympy/integrals/intpoly.py:335]: And maybe a bit too long for a single line...

Comment (faze-geek) [reply sympy/integrals/intpoly.py]: Thanks for the suggestions @oscargus  ,I tried to mimic the earlier format/style of code .For the first time I've committed suggestions directly from here (without changing my branch) , hope that it works fine !

File: sympy/integrals/intpoly.py
Comment (smichr) [inline sympy/integrals/intpoly.py:106]:
```suggestion
                _ = decompose(exp)
                if len(_) == 1 and not _.popitem()[0]:
```

File: sympy/integrals/intpoly.py
Comment (smichr) [inline sympy/integrals/intpoly.py:328]:
```suggestion
        dec_e = decompose(expr)
        if not isinstance(dec_e, list):
            return _polynomial_integrate(dec_e, facets, hp_params)
        else:
            return {exp: _polynomial_integrate(dec_e, facets, hp_params)
                for exp in expr}
```

Comment (faze-geek) [reply sympy/integrals/intpoly.py:329]: @smichr I guess your idea is of executing `decompose` only once but here it would treat a list of expressions as a single expression (which was the correction this pr was making in the first case ) .Also if we change `decompose` to check for list and solve for each expression of list internally ,then we cannot solve for polynomials (we can but with added difficulty) .We would need something to differentiate between decompose( [ 1 , x + 3 , x^2 ] ) and decompose( [ 1 , x  , 3 + x^2 ] )  .

Comment (faze-geek) [reply sympy/integrals/intpoly.py:329]: I think the current code with suggestions by @oscargus is fine (even timewise it runs in almost same time as it takes on master if `max_degree` is given as an input) . I'll resolve this conversation in some time if you don't have any particular doubts .Thanks !

**Comment by faze-geek** [comment]
@oscarbenjamin Would appreciate any comments on this pr too .Similar issues for 3-D polygons like cube , cuboid exist ( this pr corrects 2-D relevant code). I intend to fix that in a future pr once this is approved / merged.

File: sympy/integrals/intpoly.py
Comment (smichr) [inline sympy/integrals/intpoly.py:105]: how do you know you have a list for `expr` at this point?

Also, please use `e` instead of `exp` (which looks like `exp`-function, e.g. `exp(3)`)

File: sympy/integrals/intpoly.py
Comment (smichr) [inline sympy/integrals/intpoly.py:131]: this could be fixed by putting a space before the closing single quote

Comment (faze-geek) [reply sympy/integrals/intpoly.py:131]: Yeah I could do that but still the message text would need to be changed since `" Input expression be must be a valid SymPy expression "` doesn't sound right . Is the current message in my diff not correct ?

File: sympy/integrals/intpoly.py
Comment (faze-geek) [reply sympy/integrals/intpoly.py:105]: What I understood was that either there can be a single expression as input which will be handled by` max_degree=None ` condition, or there could be a list .On master the first condition check for `max_degree is not None ` case is ` if not isinstance(expr, list) and expr is not None:` .This , along with the docstring/ online documentation made me assume that list input is expected whenever `max_degree` is not None

Comment (faze-geek) [reply sympy/integrals/intpoly.py:105]: I will make the variable name changes you requested .Thanks @smichr !

File: sympy/integrals/intpoly.py
Comment (faze-geek) [reply sympy/integrals/intpoly.py:131]: I think the error message is appropriate .I have made it uniform throughout `intpoly.py` now

File: sympy/integrals/intpoly.py
Comment (faze-geek) [reply sympy/integrals/intpoly.py]: I think once this is resolved ,this pr should be ready to go in .Would appreciate any relevant comments .Thanks!

**Comment by faze-geek** [comment]
@smichr Could you check it out now ? Thanks !

**Review by certik — APPROVED** [review]
The change seems fine with me!

**Comment by faze-geek** [comment]
I believe this pr is near to going in( already approved ) .I would be glad if someone can check it out and merge .Thanks !

**Comment by certik** [comment]
@smichr, @oscarbenjamin would you mind reviewing & merging?

**Comment by faze-geek** [comment]
Thanks for the assistance @certik .I plan to make similar changes in 3D polygon codes ( have almost same issues as I fix in 2D polygons through this pr ) in a future pr .Also online documentation on intpoly is very thin for which l'll open a documentation related issue for someone to work on !

**Comment by faze-geek** [comment]
@smichr You have reviewed/merged a couple of PR's related to `intpoly` previously .I believe no more work is required here .I'll be glad if  you could check out and merge ?

**Comment by ArifAhmed1995** [comment]
> @oscarbenjamin Would appreciate any comments on this pr too .Similar issues for 3-D polygons like cube , cuboid exist ( this pr corrects 2-D relevant code). I intend to fix that in a future pr once this is approved / merged.

Hi @faze-geek ,

Apologize for the late reply. Things have been off for me lately.

It has been a long time since I worked on this. I would still suggest you make changes which are dimension independent(if possible), instead of sequentially fixing them in future PRs. If not, or if you simply wish to fix them in sequence :)
go right ahead.

You can also work on issue [#13793](https://github.com/sympy/sympy/pull/13793) if it piques your interest.

**Comment by oscarbenjamin** [comment]
I've never looked at this module before and I don't really understand what the point of it is or what these changes do but I'm happy to merge this if others think that it looks good.

**Comment by ArifAhmed1995** [comment]
This module is an implementation of the algorithm mentioned in [this paper](http://dilbert.engr.ucdavis.edu/~suku/quadrature/cls-integration.pdf)

**Comment by faze-geek** [comment]
I personally don't see any suggestions/improvements remaining here . Can someone check and merge .

**Comment by faze-geek** [comment]
Thanks to all the people who reviewed this PR .Some of you guys weren't well aware of the code and it's use still I got the help I needed which makes me appreciate the community even more :))

**Pull request merged**
